### PR TITLE
Allegro Flare as a shared lib.

### PR DESCRIPTION
**Pull request by aviallon**
TODO:

- [x] Fix linking errors
- [x] Have better dynamic library name

### CODE REVIEW

File: src/AllegroFlare/Color.cpp [hunk 1]
```diff
@@ -66,7 +66,7 @@ namespace AllegroFlare
 
 
 
-   inline ALLEGRO_COLOR operator*(float lhs, const ALLEGRO_COLOR& rhs)
+   ALLEGRO_COLOR operator*(float lhs, const ALLEGRO_COLOR& rhs)
```
Comment (MarkOates) [inline src/AllegroFlare/Color.cpp]: Great, thank you!  This one had me snagged.

I'm going to put this fix and the tweaks in the other two files into `master` now, so this PR can focus on the build.  (Also so the lib will actually build properly).

Comment (MarkOates) [reply src/AllegroFlare/Color.cpp]: ✅ K, that's added here https://github.com/allegroflare/allegro_flare/commit/ac857a0458bcbccce7f815ea67cc4614ab555aeb

**Comment by aviallon** [comment]
Also, is the "build" bash script needed ? We might want to remove it or at least rename it to avoid confusing people.

**Comment by aviallon** [comment]
Phew, I made some git mess, but I hopefully fixed it ^^

**Comment by aviallon** [comment]
@MarkOates you can merge it now. I also added CMake.

**Comment by aviallon** [comment]
I put a bit too much things in this pull request... I actually should have made a separate CMake pull request...

**Comment by MarkOates** [comment]
I've pulled the branch in locally and am trying the following steps to compile:

In the allegro_flare directory:

```
mv build build__
mkdir build/
cd build
cmake .. -D CMAKE_CXX_FLAGS="-std=gnu++11"
make
```

I'm getting a simple error of not finding allegro headers:
```
/Users/markoates/Repos/allegro_flare/include/AllegroFlare/AllegroColorAttributeDatatype.hpp:5:10: fatal error: 'allegro5/allegro_color.h' file not found
#include <allegro5/allegro_color.h>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~
```

I tried adding the `-I` to the `CMAKE_CXX_FLAGS`, but that felt naive (and also didn't work).

The `-std=gnu++11` flag was needed to prevent compile errors like:
```
/Users/markoates/Repos/allegro_flare/include/AllegroFlare/Achievements.hpp:14:58: error: a space is required between consecutive right angle
      brackets (use '> >')
      std::map<std::string, std::pair<Achievement *, bool>> all_achievements;
                                                         ^~
                                                         > >
```

Two questions:
1) What's the expected procedure for setting include directories during compile?
2) Is there a way to automatically include flags like `-std=gnu++11` so that it's not required to be passed as flags during `cmake`? (`-std=gnu++11` will likely be replaced with something more modern, eventually)

Also, I'm very new to actually using cmake as an actual development tool.  It will take me some time to ramp up.

**Comment by MarkOates** [comment]
> Also, is the "build" bash script needed ? We might want to remove it or at least rename it to avoid confusing people.

The `build` script is definitely something I use regularly, but its naming is not set.  Right now it's called [in the `quintessences` target](https://github.com/allegroflare/allegro_flare/blob/5ac6bf3bf1e9f43fc9ef88743234bb4efe68890f/Makefile.new#L83) of `Makefile.new`.   The directory `build/` would take precedence here in the collision, no doubt. The only question is what to do with `build`. 🤔

A couple options I'm thinking about:

- Could probably rename it to `qbuild` or something like that.
- Its functionality could be injected into the `Makefile.new` so it's not an isolated script.

**Comment by MarkOates** [comment]
Regarding removing `build`, I tried doing "Its functionality could be injected into the Makefile.new so it's not an isolated script" and it actually worked out quite nicely.

I just now added that to `master` in https://github.com/allegroflare/allegro_flare/commit/98d2cc858461907045b64245d508facccaad6046, which includes the removal of the `build` script. ✅

File: src/allegro_flare/drawing_interface_svg.cpp [hunk 2]
```diff
@@ -27,7 +27,8 @@ namespace allegro_flare
    {}
 
 
-
+    // Has to be defined even if pure virtual, see https://stackoverflow.com/a/15113814
```
Comment (MarkOates) [inline src/allegro_flare/drawing_interface_svg.cpp]: Can you remove this comment?  I think there were a few other places pure virtuals are defined and the extra comments aren't needed.

File: src/allegro_flare/drawing_interface_html_canvas.cpp [hunk 3]
```diff
@@ -24,6 +24,9 @@ namespace allegro_flare
    DrawingInterfaceHTMLCanvas::DrawingInterfaceHTMLCanvas()
       : DrawingInterface("html5 canvas")
    {}
+   
+   // Has to be defined even if pure virtual, see https://stackoverflow.com/a/15113814
```
Comment (MarkOates) [inline src/allegro_flare/drawing_interface_html_canvas.cpp]: ^^ Same here, regarding [my comment](https://github.com/allegroflare/allegro_flare/pull/165/files#r339318673).

File: src/allegro_flare/platform/linux/clipboard_linux.cpp [hunk 4]
```diff
@@ -0,0 +1,52 @@
+
```
Comment (MarkOates) [inline src/allegro_flare/platform/linux/clipboard_linux.cpp]: Yes, can you submit `clipboard_linux` and the renaming of `platform_generic` as a separate PR?  As it looks I would merge it right away.

Comment (aviallon) [reply src/allegro_flare/platform/linux/clipboard_linux.cpp]: Yep, I'll do that. I will also actually implement a shared clipboard for Linux (after delving in some Xorg and Wayland doc)

File: src/allegro_flare/drawing_interface_svg.cpp (hunk 2 above)
Comment (aviallon) [reply src/allegro_flare/drawing_interface_svg.cpp]: Well, this case is very special. It is needed ONLY because we are building a shared lib.

**Comment by aviallon** [comment]
@MarkOates What do you think of a CMake build, just to know ? Have you used CMake before ?

File: src/allegro_flare/platform/linux/clipboard_linux.cpp (hunk 4 above)
Comment (MarkOates) [reply src/allegro_flare/platform/linux/clipboard_linux.cpp]: > I will also actually implement a shared clipboard for Linux (after delving in some Xorg and Wayland doc)

Very cool.  I was looking into https://github.com/dacap/clip or another 3rd-party solution as a way to support all the different implementations.  I didn't get too far, but maybe something similar could work for the underlying grunt work.

**Comment by MarkOates** [comment]
> @MarkOates What do you think of a CMake build, just to know ? Have you used CMake before ?

I'm actually really excited about it!  I've not used CMake in any projects before, so we may have to support multiple build strategies for a while as I ramp up to it and have a chance to learn it.

I've only been using `Makefiles` for the time being because I haven't wanted to dive too deeply into learning a different system - I would rather spend time working on coding up cool stuff!  But `make` has time and time again given me headaches and I've also heard time and time again that `cmake` is much better.  Also, just about every project I've submitted for hackdays usually required someone else to modify or create a new `Makefile`.  I'm far from having a firm grasp on an ideal build pipeline, but it will be something important to get right when Allegro Planet supports project builds for games.

Like I said, I'm really exited to finally migrate out, but I'm brand new to configuring/using `cmake` as a build tool.  I've only used it as a few times as a build step for other projects like Allegro 5.

**Comment by aviallon** [comment]
Well, just so you know, if it is used by lots of big projects, and _especially_ big projects, there is a reason : it is extremely portable and easy to configure, and it handles a lot of compatibility stuff by itself.
If you can, I would really like to spend an hour or two while you are active to properly configure it so that it builds on any MacOS and Linux. I also do have an iMac at disposal, so I will try a few things on it too.

**Comment by MarkOates** [comment]
> If you can, I would really like to spend an hour or two while you are active to properly configure it so that it builds on any MacOS and Linux.

Are you on Slack?  There's an Allegro Planet workspace at http://allegroplanet.slack.com/, which also includes an `#allegro-flare-development` channel.  Note that I just created the channel about 2 weeks ago and there's been no activity in there yet.  I'm not going to be online there for awhile but that's a good place for quick messaging.

**Comment by MarkOates** [comment]
> Also, which version of GCC are you using ?

I'm actually using `clang` (but it aliases to `gcc` and is 100% compatible).  Exact version is:

```
Apple clang version 11.0.0 (clang-1100.0.33.8)
Target: x86_64-apple-darwin19.0.0
```

```
I'm only building using --std=c++11
```
That should work.  I wouldn't be against bumping up to `c++17` or anything more modern.

**Comment by aviallon** [comment]
Well, just so you know, if it is used by lots of big projects, and _especially_ big projects, there is a reason : it is extremely portable and easy to configure, and it handles a lot of compatibility stuff by itself.
If you can, I would really like to spend an hour or two while you are active to properly configure it so that it builds on any MacOS and Linux. I also do have an iMac at disposal, so I will try a few things on it too.

**Comment by aviallon** [comment]
> > If you can, I would really like to spend an hour or two while you are active to properly configure it so that it builds on any MacOS and Linux.
> 
> Are you on Slack? There's an Allegro Planet workspace at http://allegroplanet.slack.com/, which also includes an `#allegro-flare-development` channel. Note that I just created the channel about 2 weeks ago and there's been no activity in there yet. I'm not going to be online there today for awhile but that's a good place for quick messaging.

Yes I am. But I will have to be invited, no ?

I agree for the C++17 part.
Sorry for late replies, my laptop brutally ran out of battery @MarkOates

**Comment by sukhchainn** [comment]
I would like to be invited too.
I wanted to contribute but don't know where to start.

**Comment by aviallon** [comment]
@sukhchainn you can contribute by making a fork and creating small patches. Perhaps you could try to fix one of the issues @MarkOates created.

Also, @MarkOates, I just came back from vacations, so I'll look again into Allegro Flare, although I did a lot of work on my own concept of a C++ Allegro (see [Allegro++](https://github.com/aviallon/AllegroPP) )

**Comment by MarkOates** [comment]
> I wanted to contribute but don't know where to start.

@sukhchainn, Honestly I think one of the greatest ways you could contribute - especially if you're not sure where to start - is to just try playing around with the library and see if you can make some cool stuff!

That would really be the best way to help me figure out what patterns are working, what works well and what doesn't, what demand is for features, and all that good stuff.  It would also be good to know what features might be confusing, and where we can fill that gap by improving the API or prioritizing some much-needed documentation!

If you want to contribute to the repo itself, you could (like @aviallon mentioned) add discussion or take on any of the [issues](https://github.com/allegroflare/allegro_flare/issues) that are currently in place.  I don't think they would be as fun as playing with the library, though, and it might be better to become more familiar library first before altering the codebase.  That would be really helpful, too!

> so I'll look again into Allegro Flare, although I did a lot of work on my own concept of a C++ Allegro (see Allegro++ )

No pressure to work on AllegroFlare!  Whatever you find interesting I hope you can pursue.  It's really cool that you worked on Allegro++, I'm interested to see how you put things together! :)

**Comment by MarkOates** [comment]
Also, @aviallon and @sukhchainn, I've sent you both invites to the Allegro Planet Slack workspace.  I sent the invites to an************n@gmail.com and ss******2@gmail.com.

**Comment by MarkOates** [comment]
Cleaning up some dated branches. May investigate CMake as a build option in the future.

**Pull request closed without merging**